File: BTree.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        
class BTree:
    def __init__(self, root):
        self.root = root 
        
    def inorder_iter(self, res):
        """
        inorder tree traversal using stack
        store traversal elements in "res"
        """
        if self.root:
            stack = []
            while True:
                while self.root:
                    stack.append(self.root)
                    self.root = self.root.left
                if not stack:
                    break
                node = stack.pop()
                print(node.val)
                res.append(node.val)
                self.root = node.right
    
    def preorder_iter(self, res):
        if self.root:
            stack = [self.root]
            while stack:
                node = stack.pop()
                res.append(node.val)
                print(node.val)
                if node.right:
                    stack.append(node.right)
                if node.left:
                    stack.append(node.left)
                
    
    def postorder_iter(self, res):
        if self.root:
            stack = [self.root]
            while stack:
                node = stack.pop()
                res.append(node.val)
                if node.left:
                    stack.append(node.left)
                if node.right:
                    stack.append(node.right)
            res = res[::-1]
            print(res)
    
    def inorder_recur(self, root, res):
        if root:
            self.inorder_recur(root.left, res)
            print(root.val)
            res.append(root.val)
            self.inorder_recur(root.right, res)
    
    def preorder_recur(self, root, res):
        if root:
            print(root.val)
            res.append(root.val)
            self.preorder_recur(root.left, res)
            self.preorder_recur(root.right, res)
    
    def postorder_recur(self, root, res):
        if root:
            self.postorder_recur(root.left, res)
            self.postorder_recur(root.right, res)
            print(root.val)
            res.append(root.val)


if __name__ == "__main__":
    root = Node(7)
    n1 = Node(12)
    n2 = Node(9)
    n3 = Node(33)
    n4 = Node(91)
    n5 = Node(7)
    n6 = Node(5)
    root.left = n1
    root.right = n2
    n1.left = n3
    n3.right = n6
    n2.left = n4
    n2.right = n5
    bt = BTree(root)
    res = []
    # inorder iterative
    # bt.inorder_iter(res)
    
    # preorder iterative
    # bt.preorder_iter(res)
    
    # postorder iterative
    # bt.postorder_iter(res) 
    
    # inorder recursive
    # bt.inorder_recur(root, res) 
    
    # preorder recursive
    # bt.preorder_recur(root, res) 
    
    # postorder recursive
    bt.postorder_recur(root, res)

File: BTree.cpp
#include <iostream>
#include <vector>
#include <algorithm>

//build tree node
struct Node
{
    int val;
    Node* left;
    Node* right;
    Node(int _v) : val(_v), left(nullptr), right(nullptr) {};
};



class BTree
{
    /*
    1. construct a root pointer point to the root of a tree
    2. Three binary tree traversal has been developed using iterative methods: inorder, preorder, and postorder
    3. all iterative methods are using stack data structure
    */
    private:
        Node* root = nullptr;
    public:
        BTree(Node* rnode){root = rnode;} //class constructor to assign a binary tree to root pointer
        ~BTree(){} //destructor
        std::vector<int> inorder_iter(std::vector<int> &res)
        {
            Node* dummy = root;
            std::vector<Node*> astack;
            while(true)
            {
                while(dummy != nullptr)
                {
                    astack.push_back(dummy);
                    dummy = dummy->left;
                }
                if(astack.empty())
                    break;
                Node* temp_node = astack.back();
                astack.pop_back();
                res.push_back(temp_node->val);
                dummy = temp_node->right;
            }
            return res;
        }
        std::vector<int> preorder_iter(std::vector<int> &res)
        {
            Node* dummy = root;
            if(dummy != nullptr)
            {
                std::vector<Node*> stack;
                stack.push_back(dummy);
                while(!stack.empty())
                {
                    Node* temp_node = stack.back();
                    stack.pop_back();
                    res.push_back(temp_node->val);
                    if(temp_node->right != nullptr) 
                        stack.push_back(temp_node->right);
                    if(temp_node->left != nullptr)
                        stack.push_back(temp_node->left);
                }
            }
            return res;
        }
        std::vector<int> postorder_iter(std::vector<int> &res)
        {
            Node* dummy = root;
            if(dummy != nullptr)
            {
                std::vector<Node*> stack;
                stack.push_back(dummy);
                while(!stack.empty())
                {
                    Node* temp_node = stack.back();
                    stack.pop_back();
                    res.push_back(temp_node->val);
                    if(temp_node->left != nullptr)
                        stack.push_back(temp_node->left);
                    if(temp_node->right != nullptr)
                        stack.push_back(temp_node->right);
                }
                std::reverse(res.begin(), res.end());
            }
            return res;
        }
        
};


int main()
{
    Node* n1 = new Node(12);
    Node* n2 = new Node(33);
    Node* n3 = new Node(71);
    Node* n4 = new Node(6);
    Node* n5 = new Node(59);
    Node* n6 = new Node(17);
    Node* n7 = new Node(11);
    n1->left = n2;
    n1->right = n3;
    n2->right = n4;
    n4->left = n5;
    n3->left = n6;
    n3->right = n7;
    BTree root(n1);
    std::vector<int> res;
    root.inorder_iter(res);
    for(auto val: res)
        std::cout << val << std::endl;
    std::vector<int> res1;
    root.preorder_iter(res1);
    for(auto val1: res1)
        std::cout << val1 << std::endl;
    std::vector<int> res2;
    root.postorder_iter(res2);
    for(auto val2: res2)
        std::cout << val2 << std::endl;
    return 0;
}